Notes/BeetlebergsDesign.md

```diff
@@ -0,0 +1,377 @@
+Ranged Attacks and Trap Attacks will essentially spawn their own `Projectile` and `Trap` game objects that will have to
+have `on_update` methods to handle their behavior and thus will likely inherit from the Arcade `Sprite` class.
+
+Melee Attack `Ability` classes will an `on_update` that checks for collision with another beetle and causes damage if
+any are correctly collided. There may be a cooldown for Melee attacks where it will not immediately cause damage.
+
+Triggered Response `Ability` classes will have an `on_update` that checks if it should be triggered, and then enacts the
+effect if it should be.
+
+There are also two special `Ability` classes, the Queen's ability to spawn larvae, which can be thought of as a special
+Trap Attack, and the `Cannibal` class which allows the beetles with it to devour larvae or dead beetles to regain
+energy, which can be thought of as a special Triggered Response `Ability`.
+
+Implementing `Trait`s and `Ability`s as small shared classes follows the
+[Flyweight design pattern](https://gameprogrammingpatterns.com/flyweight.html), and should be highly performant and
+make for very self-contained code.
+
+### MVP Trait List
+
+Note that rather than nailing down numbers now, we should decide this based on experimentation, so lots of specific
+numbers have been replaced with variables like "X"
+
+* Team Composition Traits
+  * Swarm: Adds X new Basic beetles to the team
+  * Elite: Adds new Elite beetle to the team
+  * Queen: Adds new Queen beetle to the team
+  * Soldier: Adds a single new Elite beetle to the team, starting with the Mandibles Trait and an attack bonus
+  * Metamorphosis: Replaces one or more Basic beetles with Elite beetles
+* Stat Changes
+  * Dodge: Gives beetles a chance to dodge incoming projectiles based on their `awareness`. Additional gains make
+    dodging more likely
+  * Swift: Increases `max_forward_speed`
+  * Small: Decreases `size`
+  * Gigantism: Increases `size`
+  * Quick-turn: Increases `max_rotation_speed`
+  * Sniper: Increases `accuracy`
+  * Higher Perception: Increases `awareness`
+  * Armor: Increases `toughness` of Advanced beetles
+* Attack Boosts
+  * Primary ++: Increases `number_of_attacks` for primary attack
+  * Secondary ++: Increases `number_of_attacks` for secondary attack
+  * Primary Power: Increases `number_of_attacks` for primary attack
+  * Secondary Power: Increases `number_of_attacks` for secondary attack
+  * Fire Rate: Increases `fire_rate` for all attacks
+  * Farsight: Increases `range` for all Ranged attacks
+* Unique Attacks/Abilities
+  * Peashooter: basic primary attack, fires a small bullet in the current facing of the beetle
+  * Fecund: Only for Queen beetles, Trap attack that spawns larvae that eventually become Basic beetles
+  * Cannibal: Triggered Response that fires when the beetle collides with beetle corpses or larvae, consuming it,
+    healing HP and increasing size
+* Secondary Attacks/Abilities
```

Comment:
I think all the ones listed under 302 would be Elite only.

Reply:
Seems good to me, I added a little "(all Elite only)". I also went back and added mention that `Trait`s and `Ability`s should have an `acceptable_types` attribute. This will probably be a `Flag` enum, which you can read about [here in the Python documentation](https://docs.python.org/3/library/enum.html#enum.Flag).